go/mysql/collations/charset/convert.go

```diff
@@ -126,6 +128,78 @@ func Convert(dst []byte, dstCharset Charset, src []byte, srcCharset Charset) ([]
 	}
 }
 
+func Expand(dst []rune, src []byte, srcCharset Charset) []rune {
+	switch srcCharset := srcCharset.(type) {
+	case Charset_utf8mb3, Charset_utf8mb4:
+		if dst == nil {
+			return []rune(string(src))
+		}
+		dst = make([]rune, 0, len(src))
+		for _, cp := range string(src) {
+			dst = append(dst, cp)
+		}
+		return dst
+	case Charset_binary:
+		if dst == nil {
+			dst = make([]rune, 0, len(src))
+		}
+		for _, c := range src {
+			dst = append(dst, rune(c))
+		}
+		return dst
+	default:
+		if dst == nil {
+			dst = make([]rune, 0, len(src))
+		}
+		for len(src) > 0 {
+			cp, width := srcCharset.DecodeRune(src)
+			src = src[width:]
+			dst = append(dst, cp)
+		}
+		return dst
+	}
+}
+
+func Collapse(dst []byte, src []rune, dstCharset Charset) []byte {
+	switch dstCharset := dstCharset.(type) {
+	case Charset_utf8mb3, Charset_utf8mb4:
+		if dst == nil {
+			return hack.StringBytes(string(src))
+		}
+		return append(dst, hack.StringBytes(string(src))...)
+	case Charset_binary:
+		if dst == nil {
+			dst = make([]byte, 0, len(src))
+		}
+		for _, b := range src {
+			dst = append(dst, byte(b))
+		}
+		return dst
+	default:
+		nDst := 0
+		if dst == nil {
+			dst = make([]byte, len(src)*dstCharset.MaxWidth())
+		} else {
+			dst = dst[:cap(dst)]
+		}
+		for _, c := range src {
+			if len(dst)-nDst < 4 {
+				newDst := make([]byte, len(dst)*2)
+				copy(newDst, dst[:nDst])
+				dst = newDst
+			}
+			w := dstCharset.EncodeRune(dst[nDst:], c)
+			if w < 0 {
+				if w = dstCharset.EncodeRune(dst[nDst:], '?'); w < 0 {
+					break
+				}
+			}
```

Comment:
nit: you're inlining the assignment and the condition on the same line in the inner `if`, you could do the same on the outer if

Reply:
@frouioui I don't think that works since it would change what `w` means?